prov/efa/src/rdm/rxr_pkt_type_base.c

```diff
@@ -122,24 +123,35 @@ int rxr_pkt_init_data_from_op_entry(struct rxr_ep *ep,
 
 	rxr_locate_iov_pos(op_entry->iov, op_entry->iov_count, tx_data_offset,
 			   &tx_iov_index, &tx_iov_offset);
-	desc = op_entry->desc[0];
+	iov_mr = op_entry->desc[tx_iov_index];
 	assert(tx_iov_index < op_entry->iov_count);
 	assert(tx_iov_offset < op_entry->iov[tx_iov_index].iov_len);
 
-	ret = rxr_ep_use_p2p(ep, desc);
-	if (ret == 0)
-		goto copy;
+	if (pkt_entry->mr) {
+		/* When using EFA device, EFA device can access memory that
+		 * that has been registered, and p2p is allowed to be used.
+		 */
+		if (iov_mr) {
+			ret = rxr_ep_use_p2p(ep, iov_mr);
+			if (ret < 0)
+				return ret;
+			iov_accessible_by_device = ret;
+		} else {
+			iov_accessible_by_device = false;
+		}
+	} else {
+		/* When using shm, shm can access host memory without registration */
+		iov_accessible_by_device = !iov_mr || iov_mr->peer.iface == FI_HMEM_SYSTEM;
```

Comment:
Could you explain why `iov_accessible_by_device` should  be `true` if `iov_mr == NULL`?

Reply:
As comment said, shm can direct access memory without a  descriptor.

Reply:
In what case shm path needs a copy? non-p2p FI_HMEM?

Reply:
shm copy need a copy if buffer is on hmem. which means user provided a desc, and iface is not FI_HMEM_SYSTEM

Reply:
IIRC cuda to cuda via shm + ipc does not require a copy for tx, is this still needed?

Reply:
With current protocol selection, when this function is called, the memory is also on host memory.

Still, logically we should perform a copy for shm + HMEM, so I made the change.

Reply:
I see, this function is not actually called in long read protocol. Currently for shm + cuda we are using long read only, so this copy actually will not happen. If we are going to use eager for shm + cuda for small msg, this will be called. Make sense.

Reply:
`rxr_pkt_init_data_from_op_entry` is called by `rxr_pkt_init_rtm`. But `rxr_pkt_init_rtm` is only called by protocols that have data in the rtm packet. I think the name `rxr_pkt_init_rtm` is a bit misleading though because not all rtm really use it. But it can be a refactor in the future.